x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/analytics/process/AnalyticsResultProcessor.java

```diff
@@ -0,0 +1,115 @@
+
+    private final Client client;
+    private final DataFrameDataExtractor dataExtractor;
+    private List<DataFrameDataExtractor.Row> currentDataFrameRows;
+    private List<AnalyticsResult> currentResults;
+    private final CountDownLatch completionLatch = new CountDownLatch(1);
+
+    public AnalyticsResultProcessor(Client client, DataFrameDataExtractor dataExtractor) {
+        this.client = Objects.requireNonNull(client);
+        this.dataExtractor = Objects.requireNonNull(dataExtractor);
+    }
+
+    public void awaitForCompletion() {
+        try {
+            if (completionLatch.await(30, TimeUnit.MINUTES) == false) {
+                LOGGER.warn("Timeout waiting for results processor to complete");
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            LOGGER.info("Interrupted waiting for results processor to complete");
+        }
+    }
+
+    public void process(AnalyticsProcess process) {
+
+        try {
+            Iterator<AnalyticsResult> iterator = process.readAnalyticsResults();
+            while (iterator.hasNext()) {
+                try {
+                    AnalyticsResult result = iterator.next();
+                    if (dataExtractor.hasNext() == false) {
+                        return;
+                    }
+                    if (currentDataFrameRows == null) {
+                        Optional<List<DataFrameDataExtractor.Row>> nextBatch = dataExtractor.next();
+                        if (nextBatch.isPresent() == false) {
+                            return;
+                        }
+                        currentDataFrameRows = nextBatch.get();
+                        currentResults = new ArrayList<>(currentDataFrameRows.size());
+                    }
+                    currentResults.add(result);
+                    if (currentResults.size() == currentDataFrameRows.size()) {
+                        joinCurrentResults();
+                        currentDataFrameRows = null;
+                    }
+                } catch (Exception e) {
+                    LOGGER.warn("Error processing analytics result", e);
+                }
+
+            }
+        } catch (Exception e) {
+            LOGGER.error("Error parsing analytics output", e);
+        } finally {
+            completionLatch.countDown();
+            process.consumeAndCloseOutputStream();
+        }
+    }
+
+    private void joinCurrentResults() {
+        BulkRequest bulkRequest = new BulkRequest();
+        for (int i = 0; i < currentDataFrameRows.size(); i++) {
+            DataFrameDataExtractor.Row row = currentDataFrameRows.get(i);
+            if (row.shouldSkip()) {
+                continue;
+            }
+            AnalyticsResult result = currentResults.get(i);
+            SearchHit hit = row.getHit();
+            Map<String, Object> source = new HashMap(hit.getSourceAsMap());
```

Comment:
Do we want the new fields to come at the end of the source document?  If so, this should use `LinkedHashMap` instead of `HashMap`.  (Maybe we don't care though.)

Reply:
Good question. I am not sure what is actually affected by the order. But I think it makes sense to append them at the end as the least invasive annotation of the original rows.